networkx/algorithms/non_randomness.py

```diff
@@ -12,86 +12,150 @@
 
+    When the number of communities `k` is not specified,
+    :func:`~networkx.algorithms.community.label_propagation_communities`
+    is used to compute it.
+    This algorithm can give different results depending on
+    the order of nodes and edges in the graph.
+    For example, while the following graphs are identical,
+    computing the non-randomness of each of them yields different results:
+
+    >>> G1, G2 = nx.Graph(), nx.Graph()
+    >>> G1.add_edges_from([(0, 1), (1, 2), (1, 3), (3, 4)])
+    >>> G2.add_edges_from([(0, 1), (1, 3), (1, 2), (3, 4)])
+    >>> [round(r, 6) for r in nx.non_randomness(G1)]
+    [-1.847759, -5.842437]
+    >>> [round(r, 6) for r in nx.non_randomness(G2)]
+    Traceback (most recent call last):
+     ...
+    ValueError: invalid number of communities for graph with 5 nodes and 4 edges: 2
+
+    This is because the community detection algorithm finds
+    1 community in ``G1`` and 2 communities in ``G2``.
+    This can be resolved by specifying the number of communities `k`:
+
+    >>> [round(r, 6) for r in nx.non_randomness(G2, k=1)]
+    [-1.847759, -5.842437]
+
     Notes
     -----
-    This computes Eq. (4.4) and (4.5) in Ref. [1]_.
-
-    If a weight field is passed, this algorithm will use the eigenvalues
-    of the weighted adjacency matrix to compute Eq. (4.4) and (4.5).
+    If a `weight` field is passed, this algorithm will use the eigenvalues
+    of the weighted adjacency matrix instead.
+
+    The output of this function corresponds to (4.4) and (4.5) in [1]_.
+    A lower value of $R^*_G$ indicates a more random graph;
+    one can think of $1 - \\Phi(R_G^*)$ as the similarity
+    between the graph and a random graph,
+    where $\\Phi(x)$ is the cumulative distribution function
+    of the standard normal distribution.
+
+    Theorem 2 in [2]_ states that for any graph $G$
+    with $n$ nodes, $m$ edges, and $k$ communities,
+    its non-randomness is bounded below by the non-randomness of an
+    $r$-regular graph (a graph where each node has degree $r$),
+    and bounded above by the non-randomness of an $l$-complete graph
+    (a graph where each community is a clique of $l$ nodes).
 
     References
     ----------
     .. [1] Xiaowei Ying and Xintao Wu,
            On Randomness Measures for Social Networks,
            SIAM International Conference on Data Mining. 2009
+           https://doi.org/10.1137/1.9781611972795.61
+    .. [2] Ying, Xiaowei & Wu, Leting & Wu, Xintao. (2012).
+           A Spectrum-Based Framework for Quantifying Randomness of Social Networks.
+           IEEE Transactions on Knowledge and Data Engineering 23(12):1842--1856.
+           https://dl.acm.org/doi/abs/10.1109/TKDE.2010.218
     """
     import numpy as np
 
     # corner case: graph has no edges
     if nx.is_empty(G):
-        raise nx.NetworkXError("non_randomness not applicable to empty graphs")
+        raise nx.NetworkXError("graph has no edges")
```

Comment:
Changing exception messages is a recipe for bogging down review as it's (arguably) unnecessary code churn unrelated to the main changes.

I don't have strong feelings in this case, but generally speaking I'd recommend avoiding such changes to avoid distracting side-discussions during review. If you feel very strongly about updating an exception message, the best way to go about it (IMO) is to open a dedicated PR for it!

Reply:
You're right that it's an unnecessary change. I'm juggling two feelings here: the very strong feeling I have about the lack of standardization of error messages (and types for that matter) across the repo on one hand, and the desire to keep changes as small and "meaningful" as possible (with the added problem that changes, especially to exception types, can break things for downstream users) on the other. I'm not sure whether anyone else shares these concerns, or how to best go about alleviating them, so I appreciate your perspective!